File: packages/python/src/playbook_md/validator.py
"""
PLAYBOOK.md Semantic Validator

Performs additional semantic checks beyond what the parser does.
The parser handles syntax; the validator checks cross-references,
variable scoping, and other semantic constraints.

Ported from the TypeScript implementation in packages/core/src/validator.ts.
"""

from __future__ import annotations

import re
from typing import Optional

from .parser import parse_playbook
from .types import (
    ArtifactType,
    ParseError,
    ParseWarning,
    PlaybookDefinition,
    Step,
    ValidationResult,
)

VALID_ARTIFACT_TYPES: list[str] = [
    "markdown",
    "json",
    "mermaid",
    "chartjs",
    "html_css",
    "javascript",
    "typescript",
]

RE_INTERPOLATION = re.compile(r"\{\{(\w+)(?::\w+(?::[^}]*)?)?}}")


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _collect_interpolations(text: str) -> list[dict[str, object]]:
    """Collect all {{variable}} references from a string."""
    refs: list[dict[str, object]] = []
    for match in RE_INTERPOLATION.finditer(text):
        refs.append({"name": match.group(1), "offset": match.start()})
    return refs


def _check_interpolations(
    content: str,
    step: Step,
    input_names: set[str],
    available_outputs: set[str],
    warnings: list[ParseWarning],
) -> None:
    """Check {{variable}} interpolations in step content."""
    refs = _collect_interpolations(content)
    for ref in refs:
        name = str(ref["name"])
        if name not in input_names and name not in available_outputs:
            warnings.append(
                ParseWarning(
                    line=step.line,
                    message=f'Step {step.label} references undeclared variable "{{{{{name}}}}}"',
                )
            )


# ---------------------------------------------------------------------------
# Core validation
# ---------------------------------------------------------------------------


def _validate_definition(definition: PlaybookDefinition) -> ValidationResult:
    """Validate a parsed PlaybookDefinition for semantic correctness."""
    fatal_errors: list[ParseError] = []
    warnings: list[ParseWarning] = []

    input_names = {inp.name for inp in definition.inputs}

    # Track which @output variables are available at each point in execution.
    # We walk steps in order; an @output from step N is available in step N+1 onward.
    outputs_before: dict[int, set[str]] = {}  # step_index -> outputs available before this step
    all_outputs: set[str] = set()

    # First pass: collect outputs in execution order
    for i, step in enumerate(definition.steps):
        outputs_before[i] = set(all_outputs)

        if step.output_var:
            all_outputs.add(step.output_var)

        # Also collect outputs from branch sub-steps (they may or may not execute,
        # but the variables they declare become "potentially available" downstream)
        if step.branches:
            for branch in step.branches:
                for sub_step in branch.steps:
                    if sub_step.output_var:
                        all_outputs.add(sub_step.output_var)

    # Check 1: @output variable names must not shadow input names
    for step in definition.steps:
        if step.output_var and step.output_var in input_names:
            fatal_errors.append(
                ParseError(
                    line=step.line,
                    message=f'@output variable "{step.output_var}" in step {step.number} shadows an input name',
                )
            )

        if step.branches:
            for branch in step.branches:
                for sub_step in branch.steps:
                    if sub_step.output_var and sub_step.output_var in input_names:
                        fatal_errors.append(
                            ParseError(
                                line=sub_step.line,
                                message=f'@output variable "{sub_step.output_var}" in sub-step {sub_step.label} shadows an input name',
                            )
                        )

    # Check 2: Verify {{variable}} interpolations reference declared inputs or prior @output vars
    for i, step in enumerate(definition.steps):
        available_outputs = outputs_before[i]

        _check_interpolations(
            step.content, step, input_names, available_outputs, warnings
        )

        # Check branch sub-step content
        if step.branches:
            for branch in step.branches:
                for sub_step in branch.steps:
                    _check_interpolations(
                        sub_step.content,
                        sub_step,
                        input_names,
                        available_outputs,
                        warnings,
                    )

        # Check @tool arguments for interpolations
        if step.tool_call and step.tool_call.arguments:
            refs = _collect_interpolations(step.tool_call.arguments)
            for ref in refs:
                name = str(ref["name"])
                if name not in input_names and name not in available_outputs:
                    warnings.append(
                        ParseWarning(
                            line=step.line,
                            message=f'Step {step.number} @tool argument references undeclared variable "{{{{{name}}}}}"',
                        )
                    )

    # Check 3: Verify branch conditions reference variables that exist at that point
    for i, step in enumerate(definition.steps):
        available_outputs = outputs_before[i]

        if step.branches:
            for branch in step.branches:
                if branch.condition:
                    var_name = branch.condition.variable
                    if var_name not in input_names and var_name not in available_outputs:
                        warnings.append(
                            ParseWarning(
                                line=step.line,
                                message=f'Branch condition in step {step.number} references undeclared variable "{var_name}"',
                            )
                        )

    # Check 4: Verify artifact type is valid
    if (
        definition.artifact_type
        and definition.artifact_type not in VALID_ARTIFACT_TYPES
    ):
        warnings.append(
            ParseWarning(
                message=f'Unknown artifact type "{definition.artifact_type}". Valid types: {", ".join(VALID_ARTIFACT_TYPES)}',
            )
        )

    return ValidationResult(
        valid=len(fatal_errors) == 0,
        fatal_errors=fatal_errors,
        warnings=warnings,
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def validate_playbook(markdown: str) -> ValidationResult:
    """Parse and validate a playbook markdown string.

    Runs the parser first, then performs additional semantic validation
    on the resulting definition.

    Args:
        markdown: The raw PLAYBOOK.md markdown content.

    Returns:
        A ValidationResult indicating whether the playbook is valid,
        along with any fatal errors and warnings.
    """
    parse_result = parse_playbook(markdown)

    # If parsing failed, return parser errors as fatal errors
    if parse_result.definition is None:
        return ValidationResult(
            valid=False,
            fatal_errors=parse_result.errors,
            warnings=parse_result.warnings,
        )

    # Run semantic validation on the parsed definition
    validation_result = _validate_definition(parse_result.definition)

    # Merge parser warnings with validation warnings
    return ValidationResult(
        valid=len(validation_result.fatal_errors) == 0,
        fatal_errors=validation_result.fatal_errors,
        warnings=parse_result.warnings + validation_result.warnings,
    )

File: packages/core/src/validator.ts
/**
 * PLAYBOOK.md Semantic Validator
 *
 * Performs additional semantic checks beyond what the parser does.
 * The parser handles syntax; the validator checks cross-references,
 * variable scoping, and other semantic constraints.
 */

import { parsePlaybook } from './parser';
import type {
  ParseResult,
  ParseError,
  ParseWarning,
  PlaybookDefinition,
  Step,
  Branch,
  ArtifactType,
} from './types';

export interface ValidationResult {
  valid: boolean;
  fatal_errors: ParseError[];
  warnings: ParseWarning[];
}

const VALID_ARTIFACT_TYPES: ArtifactType[] = [
  "markdown", "json", "mermaid", "chartjs", "html_css", "javascript", "typescript",
];

const RE_INTERPOLATION = /\{\{(\w+)(?::\w+(?::[^}]*)?)?}}/g;

/**
 * Collect all {{variable}} references from a string.
 */
function collectInterpolations(text: string): { name: string; offset: number }[] {
  const refs: { name: string; offset: number }[] = [];
  let match: RegExpExecArray | null;
  const re = new RegExp(RE_INTERPOLATION.source, 'g');
  while ((match = re.exec(text)) !== null) {
    refs.push({ name: match[1], offset: match.index });
  }
  return refs;
}

/**
 * Validate a parsed PlaybookDefinition for semantic correctness.
 */
function validateDefinition(def: PlaybookDefinition): ValidationResult {
  const fatal_errors: ParseError[] = [];
  const warnings: ParseWarning[] = [];

  const inputNames = new Set(def.inputs.map(i => i.name));

  // Track which @output variables are available at each point in execution.
  // We walk steps in order; an @output from step N is available in step N+1 onward.
  const outputsBefore = new Map<number, Set<string>>(); // stepIndex -> outputs available before this step
  const allOutputs = new Set<string>();

  // First pass: collect outputs in execution order
  for (let i = 0; i < def.steps.length; i++) {
    outputsBefore.set(i, new Set(allOutputs));

    const step = def.steps[i];
    if (step.output_var) {
      allOutputs.add(step.output_var);
    }

    // Also collect outputs from branch sub-steps (they may or may not execute,
    // but the variables they declare become "potentially available" downstream)
    if (step.branches) {
      for (const branch of step.branches) {
        for (const subStep of branch.steps) {
          if (subStep.output_var) {
            allOutputs.add(subStep.output_var);
          }
        }
      }
    }
  }

  // Check 1: @output variable names must not shadow input names
  for (const step of def.steps) {
    if (step.output_var && inputNames.has(step.output_var)) {
      fatal_errors.push({
        line: step.line,
        message: `@output variable "${step.output_var}" in step ${step.number} shadows an input name`,
      });
    }

    if (step.branches) {
      for (const branch of step.branches) {
        for (const subStep of branch.steps) {
          if (subStep.output_var && inputNames.has(subStep.output_var)) {
            fatal_errors.push({
              line: subStep.line,
              message: `@output variable "${subStep.output_var}" in sub-step ${subStep.label} shadows an input name`,
            });
          }
        }
      }
    }
  }

  // Check 2: Verify {{variable}} interpolations reference declared inputs or prior @output vars
  for (let i = 0; i < def.steps.length; i++) {
    const step = def.steps[i];
    const availableOutputs = outputsBefore.get(i)!;

    checkInterpolations(step.content, step, inputNames, availableOutputs, warnings);

    // Check branch sub-step content
    if (step.branches) {
      for (const branch of step.branches) {
        for (const subStep of branch.steps) {
          checkInterpolations(subStep.content, subStep, inputNames, availableOutputs, warnings);
        }
      }
    }

    // Check @tool arguments for interpolations
    if (step.tool_call?.arguments) {
      const refs = collectInterpolations(step.tool_call.arguments);
      for (const ref of refs) {
        if (!inputNames.has(ref.name) && !availableOutputs.has(ref.name)) {
          warnings.push({
            line: step.line,
            message: `Step ${step.number} @tool argument references undeclared variable "{{${ref.name}}}"`,
          });
        }
      }
    }
  }

  // Check 3: Verify branch conditions reference variables that exist at that point
  for (let i = 0; i < def.steps.length; i++) {
    const step = def.steps[i];
    const availableOutputs = outputsBefore.get(i)!;

    if (step.branches) {
      for (const branch of step.branches) {
        if (branch.condition) {
          const varName = branch.condition.variable;
          if (!inputNames.has(varName) && !availableOutputs.has(varName)) {
            warnings.push({
              line: step.line,
              message: `Branch condition in step ${step.number} references undeclared variable "${varName}"`,
            });
          }
        }
      }
    }
  }

  // Check 4: Verify artifact type is valid
  if (def.artifact_type && !VALID_ARTIFACT_TYPES.includes(def.artifact_type)) {
    // Check if it's a dynamic variable reference like {{output_format}}
    const dynamicMatch = def.artifact_type.match(/^\{\{(\w+)\}\}$/);
    if (dynamicMatch) {
      // W6: check that the referenced variable is declared
      const varName = dynamicMatch[1];
      if (!inputNames.has(varName) && !allOutputs.has(varName)) {
        warnings.push({
          message: `Artifact type references undeclared variable "{{${varName}}}"`,
        });
      }
    } else {
      warnings.push({
        message: `Unknown artifact type "${def.artifact_type}". Valid types: ${VALID_ARTIFACT_TYPES.join(', ')}`,
      });
    }
  }

  return {
    valid: fatal_errors.length === 0,
    fatal_errors,
    warnings,
  };
}

/**
 * Check {{variable}} interpolations in step content.
 */
function checkInterpolations(
  content: string,
  step: Step,
  inputNames: Set<string>,
  availableOutputs: Set<string>,
  warnings: ParseWarning[],
): void {
  const refs = collectInterpolations(content);
  for (const ref of refs) {
    if (!inputNames.has(ref.name) && !availableOutputs.has(ref.name)) {
      warnings.push({
        line: step.line,
        message: `Step ${step.label} references undeclared variable "{{${ref.name}}}"`,
      });
    }
  }
}

/**
 * Parse and validate a playbook markdown string.
 *
 * Runs the parser first, then performs additional semantic validation
 * on the resulting definition.
 */
export function validatePlaybook(markdown: string): ValidationResult {
  const parseResult: ParseResult = parsePlaybook(markdown);

  // If parsing failed, return parser errors as fatal errors
  if (!parseResult.definition) {
    return {
      valid: false,
      fatal_errors: parseResult.errors,
      warnings: parseResult.warnings,
    };
  }

  // Run semantic validation on the parsed definition
  const validationResult = validateDefinition(parseResult.definition);

  // Merge parser warnings with validation warnings
  return {
    valid: validationResult.fatal_errors.length === 0,
    fatal_errors: validationResult.fatal_errors,
    warnings: [...parseResult.warnings, ...validationResult.warnings],
  };
}
